### ISSUE TITLE
`to_number` built-in function should reject "Inf", "Infinity" and "NaN" values

### ISSUE THREAD

**Issue by anderseknert** [open-policy-agent/opa#7200]
While providing these will omit errors, these errors are from JSON marshalling of the decision and not from input validation. Compare:

```
opa eval -f pretty --strict-builtin-errors 'to_number("Infinite")'
1 error occurred: to_number("Infinite"): eval_builtin_error: to_number: strconv.ParseFloat: parsing "Infinite": invalid syntax
```

```
opa eval -f pretty --strict-builtin-errors 'to_number("Infinity")'
json: invalid number literal "Infinity"
```

This since the `strcov.ParseFloat` function [accepts these values](https://pkg.go.dev/strconv#ParseFloat), but OPA obviously should not.

**Comment by sikehish** [comment]
I'm looking into this issue :)

**Comment by sikehish** [comment]
> While providing these will omit errors, these errors are from later JSON marshalling and not from input validation. Compare:
> 
> ```
> opa eval -f pretty --strict-builtin-errors 'to_number("Infinite")'
> 1 error occurred: to_number("Infinite"): eval_builtin_error: to_number: strconv.ParseFloat: parsing "Infinite": invalid syntax
> ```
> 
> ```
> opa eval -f pretty --strict-builtin-errors 'to_number("Infinity")'
> json: invalid number literal "Infinity"
> ```
> 
> This since the `strcov.ParseFloat` function [accepts these values](https://pkg.go.dev/strconv#ParseFloat), but OPA obviously should not.

Hi @anderseknert . I've made the required changes. I just wanted to know where the test cases have to be added in this case(if that's required, as I've manually tested for all three inputs: "Inf", "Infinity" and "NaN").

**Comment by anderseknert** [comment]
Hi there! 

There was a PR open on that already, but unfortunately it doesn't seem like it was linked https://github.com/open-policy-agent/opa/pull/7201 ... and also that there's quite some work left to do there. 

@raj921 , do you want to keep going there?

As for test cases (no matter who adds them, the preferred way is to add "YAML tests" under `test/cases/testdata/v1/casts` ... you can look at some existing tests to see what they look like (you can also add attributes to test for errors returned). 

The comment in [this file](https://github.com/open-policy-agent/opa/blob/976e2d467b6964973fda338ca1b154d09bd75dc4/test/cases/testdata/v0/helloworld/test-helloworld-1.yaml) contains some information for how to run those tests isolated (if not as part of the whole `make test` suite).

Sorry for having snoozed for a while, it's evening here and 2 kids are demanding my attention 😅

**Comment by raj921** [comment]
no sir I will try another issue

**Comment by anderseknert** [comment]
Alright, I'll keep that in mind, and can assign you to one when I find one good to start with 👍 Thanks a lot!

**Comment by anderseknert** [comment]
@sikehish you can proceed with adding some tests then as previously mentioned. Let me know if you have any questions!

**Comment by sikehish** [comment]
> Hi there!
> 
> There was a PR open on that already, but unfortunately it doesn't seem like it was linked #7201 ... and also that there's quite some work left to do there.
> 
> @raj921 , do you want to keep going there?
> 
> As for test cases (no matter who adds them, the preferred way is to add "YAML tests" under `test/cases/testdata/v1/casts` ... you can look at some existing tests to see what they look like (you can also add attributes to test for errors returned).
> 
> The comment in [this file](https://github.com/open-policy-agent/opa/blob/976e2d467b6964973fda338ca1b154d09bd75dc4/test/cases/testdata/v0/helloworld/test-helloworld-1.yaml) contains some information for how to run those tests isolated (if not as part of the whole `make test` suite).
> 
> Sorry for having snoozed for a while, it's evening here and 2 kids are demanding my attention 😅

No issues:) Should I proceed with adding the test cases? I thought it was up for grabs as there weren't any comments indicating that the issue was being resolved.

**Comment by sikehish** [comment]
> @sikehish you can proceed with adding some tests then as previously mentioned. Let me know if you have any questions!

Do I need to create a new file for it or can I make changes in any of the existing test case yml file?

**Comment by anderseknert** [comment]
I'd probably create a new file and name it something appropriate, but either way is fine. Don't mind the naming or the contents of the old files too much. They were generated from code long ago, and new tests don't need to follow them.

**Comment by sikehish** [comment]
> I'd probably create a new file and name it something appropriate, but either way is fine. Don't mind the naming or the contents of the old files too much. They were generated from code long ago, and new tests don't need to follow them.

Cool. Also is there a way to generate the tests or do I have to write it manually?

**Comment by anderseknert** [comment]
The YAML file? You'd write that yourself. But you can obviously copy the structure from other files.

**Comment by sikehish** [comment]
> The YAML file? You'd write that yourself. But you can obviously copy the structure from other files.

Yup, I was referring to the YAML file. The yml file names looked like they were generated so was a bit curious xD.

**Comment by sikehish** [comment]
Thanks for the opportunity @anderseknert !

**Comment by anderseknert** [comment]
Thanks for the contribution!

**Comment by sikehish** [comment]
> Thanks for the contribution!

Hi @anderseknert . Why does opa eval -f pretty --strict-builtin-errors  not work with values prefixed with +? I was getting JSON marshalling error when I tried for +42. Is this a bug?

**Comment by anderseknert** [comment]
Yeah, looks like one to me. I'd expect that to just be stripped off, or if it's meant to be an error, surfaced as a built-in error. I don't really see any point in not allowing that though. Go ahead and submit a fix if you want :)

**Comment by sikehish** [comment]
> Yeah, looks like one to me. I'd expect that to just be stripped off, or if it's meant to be an error, surfaced as a built-in error. I don't really see any point in not allowing that though. Go ahead and submit a fix if you want :)

Should I raise an issue and then start working on it?
